examples/benchmark/benchmark.py

```diff
@@ -0,0 +1,113 @@
+#  Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+#  // SPDX-License-Identifier: BSD
+import hydra
+import torchdata
+from omegaconf import DictConfig
+from s3torchconnector import S3IterableDataset, S3Reader, S3MapDataset
+from torch.utils.data import DataLoader, default_collate
+from torchdata.datapipes.utils import StreamWrapper
+
+from benchmark_utils import ResourceMonitor
+from models import Entitlement, ViT
+
+
+@hydra.main(version_base=None, config_path="configuration")
+def run_experiment(config: DictConfig):
+    if config.training.model == "entitlement":
+        model = Entitlement()
+    elif config.training.model == "vit":
+        model = ViT(1000, config.checkpoint)
+    else:
+        raise Exception(f"unknown model {config.training.model}")
+
+    dataset = make_dataset(
+        config.dataloader.kind,
+        config.dataset.sharding,
+        config.dataset.prefix_uri,
+        config.dataset.region,
+        model.load_sample,
+        num_workers=config.dataloader.num_workers,
+    )
+    dataloader = make_dataloader(
+        dataset, config.dataloader.num_workers, config.dataloader.batch_size
+    )
+
+    monitor = ResourceMonitor()
+    monitor.start()
+    result = model.train(dataloader, config.training.max_epochs)
+    result.resource_data = monitor.get_full_data()
+
+    # TODO: Decide if we need to do averaging in Monitor vs in ExperimentResult
+    result.avg_resource_data = monitor.get_avg_data()
+    monitor.stop()
+
+    # TODO: We are currently only printing the result of the experiment here. We should either write it to a file
+    # for further processing, or use CALLBACKS to actually graph/analyse the result. Ideally we should do both.
+    print(
+        f"{config.dataloader.kind} trained {config.training.model} in "
+        f"{result.training_time:.4f}s with {result.throughput:.4f} samples per second"
+    )
+
+    print(
+        f"Resource usage of the workload was as follows:\n"
+        f"{result.avg_resource_data}"
+    )
+
+
+def make_dataset(
+    kind: str,
+    sharded: bool,
+    prefix_uri: str,
+    region: str,
+    load_sample,
+    num_workers: int,
+):
+    if kind == "s3iterabledataset":
+        dataset = S3IterableDataset.from_prefix(prefix_uri, region=region)
+        dataset = torchdata.datapipes.iter.IterableWrapper(dataset)
+        if num_workers > 0:
+            dataset = dataset.sharding_filter()
+        if sharded:
+            dataset = dataset.map(tar_to_tuple)
+            dataset = dataset.load_from_tar()
+        return dataset.map(load_sample)
+    if kind == "s3mapdataset":
+        if sharded:
+            raise Exception(f"Sharding is not supported for {kind}")
```

Comment:
Should this be a `NotImplementedError`? I don't like raising bare exceptions

Reply:
Agree with the comment re bare exceptions. I don't think it will ever be supported either either no? I'm thinking a `ValueError` would be more appropriate.

Reply:
Using a `ValueError` sounds right to me